core/tools/embed_shaders.py

```
@@ -1,7 +1,20 @@
 import sys, pathlib
 
-SHADER_DIR = pathlib.Path(sys.argv[1])
-OUTPUT_FILE = pathlib.Path(sys.argv[2])
+def require_within_root(path: pathlib.Path, root: pathlib.Path, label: str) -> pathlib.Path:
+    resolved = path.resolve()
+    root_resolved = root.resolve()
+    try:
+        resolved.relative_to(root_resolved)
+    except ValueError:
+        raise ValueError(f"{label} must be within {root_resolved}: {resolved}")
```

Ruff (0.15.12)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)